### PR TITLE
Feature/samples

### PR DESCRIPTION
**Pull request by Jammy2211**
Refactoring sampls to remove many redundant methods.

### CODE REVIEW

File: autofit/non_linear/nest/dynesty/abstract.py [hunk 1]
```diff
@@ -360,7 +361,7 @@ def sampler_from(
     ):
         raise NotImplementedError()
 
-    def check_pool(self, uses_pool : bool, pool: Pool):
+    def check_pool(self, uses_pool : bool, pool: "Pool"):
```
Comment (rhayes777) [inline autofit/non_linear/nest/dynesty/abstract.py:364]: Why can't the type be imported here?

File: autofit/non_linear/samples/samples.py [hunk 2]
```diff
@@ -12,7 +14,240 @@
...
+
+    Returns
+    -------
+        A function that returns a 2D image.
+    """
+
+    @wraps(func)
+    def wrapper(
+        obj,
+        sample_index,
```
Comment (rhayes777) [inline autofit/non_linear/samples/samples.py:149]: sample_index can be implicitly contianed in args or kwargs.

Comment (rhayes777) [reply autofit/non_linear/samples/samples.py:149]: Do we really need to pass as_instance to the decorated function?

**Review by rhayes777 — APPROVED** [review]

File: autofit/non_linear/samples/samples.py (hunk 2 above)
Comment (Jammy2211) [reply autofit/non_linear/samples/samples.py:149]: Lets dicuss next week.

**Pull request merged**